### PR TITLE
Closed-book exam functionality

### PR DESCRIPTION
**Pull request by vec4**
This is all the changes I've made to the web side for adding closed-book exam functionality to FreeCC.

In summary:

- Created a wrapper around PWC's JS API so that we can create instances in the background without having to bind them on a websocket with a terminal in the browser.
- Created a layout "exam" for closed-book exams. Instead of a terminal on the right-hand side, it contains a code text editor. There is a status bar on the bottom of the editor that indicates current connection to PWC and session time remaining. There are also three buttons: build code, run code (on the test cases), and view console output.
- Compilation errors are returned in the HTTP response from PWC and shown in the console window.
- Added a parameter in the _config.yml that contains the endpoint for exam resources (like in the PWC changes I made). It needs to be changed to the AWS server running FreeCC as I am using a temporary AWS host currently. It is a one line change and the only thing that needs to be changed.

I will write a guide on how to add a new exam and polish up one of my mine I use for testing to publish to the site.

### CODE REVIEW

**Comment by chunhualiao** [comment]
There are quite some commits to review. I suggest to push them into a
branch first then ask Anjia to try it out on a test server.  We can then
merge after that.

**Comment by vec4** [comment]
Sounds good. This is all in the "online-testing" branch.

**Comment by ouankou** [comment]
I'll clone this new branch to another separate repo and deploy it to Git Pages for testing.

**Comment by vec4** [comment]
@ouankou  Let me know if you need a simple example exam for testing it out.

File: _config.yml [hunk 1]
```diff
@@ -60,5 +60,8 @@ plugins_dir:
 # Play with Docker
 pwdurl: http://lab.freecompilercamp.org
 
+# Upload endpoint for closed-book exams
+exam_endpoint: http://ec2-54-241-120-89.us-west-1.compute.amazonaws.com/llnl-freecompilercamp/freecc_exams/
```
Comment (ouankou) [inline _config.yml:64]: Currently it uses an individual server to host the files used in the exam. Maybe it's better to store those files in a public GitHub repo as well. It's easier for management and usage.

**Review by ouankou — APPROVED** [review]
I've deployed the new training training website and PWC server successfully. It can be accessed at http://classic.ouankou.com
A sample exam from Justin has been manually added to the testing server, which is not included in this pull request.

I went through the practice and exam and both work as expected. In the exam, the output from a failed building can be shown in the console. The files used in a exam are stored in a separate sever now. We may move them into a public repo.

http://classic.ouankou.com/rose-complex-types-practical/
http://classic.ouankou.com/rose-complextypes-test/

**Comment by ouankou** [comment]
@chunhualiao I've reviewed the code and tested it. Everything looks fine. May I merge them and deploy to our production server now?

**Comment by chunhualiao** [comment]
go ahead please.

File: _config.yml (hunk 1 above)
Comment (vec4) [reply _config.yml:64]: Fixed with  https://github.com/freeCompilerCamp/freecompilercamp.github.io/commit/becfa91b3c637442282a2eff83ceb003a328de09, https://github.com/freeCompilerCamp/freecompilercamp.github.io/commit/7e274078d301569f74f5ab76781f327ccb424a84, and https://github.com/freeCompilerCamp/play-with-compiler/commit/059d00a9c22765fa5d27fce7d19e6cfc119dcc09. 

Exam resources can be uploaded to the https://github.com/freeCompilerCamp/code-for-rose-tutorials repo for ROSE exams. LLVM exam support will be coming soon.

**Pull request merged**